README.md
# flutter_screen_adapter

[![pub package](https://img.shields.io/pub/v/flutter_screen_adapter.svg)](https://pub.dartlang.org/packages/flutter_screen_adapter)

[README of English](https://github.com/tal-tech/flutter_screen_adapter/blob/master/README_EN.md)

`Flutter` 应用不同终端屏幕适配方案。

<img src='https://raw.githubusercontent.com/tal-tech/flutter_screen_adapter/master/doc/example_zh.png' alt='example' />

## 使用步骤

### 步骤1：添加 `flutter_screen_adapter` 到 `pubspec.yaml` 中

```yaml
dependencies:
  flutter:
    sdk: flutter

  flutter_screen_adapter: ^0.0.1
```

### 步骤2：初始化

```dart
import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {

    /// 注意：
    /// 一定要在 `WidgetsApp` 或者 `MaterialApp widget` 中初始化
    /// 否则，会因为 `MediaQuery` 找不到报错

    ScreenAdapter.init(context);
    return ...
  }
}
```

### 步骤3：设置元素大小

```dart
// 正方形
Container(
  width: ScreenAdapter.value(100),
  height: ScreenAdapter.value(100),
  color: Colors.yellow,
  child: null
)

// 长方形

Container(
  width: ScreenAdapter.value(200),
  height: ScreenAdapter.value(100),
  color: Colors.red,
  child: null
)

```

## 提供的属性

<img src="https://raw.githubusercontent.com/tal-tech/flutter_screen_adapter/master/doc/property.png" alt="property"  >

### 这里列举两个常用的属性

```dart
...

/// 根据设备信息得到自适应值
/// 用来使用不同屏幕达到跟设计稿一致效果
///
/// [vertical] 是否是垂直自适应数值，默认是水平

static double value(num value, {vertical = false});

/// 获取当前设备屏幕宽度
///
static double get screenWidth;

...
```

CHANGELOG.md
## [0.0.2]
 - Improve documentation

## [0.0.1]
 - First version

test/flutter_screen_adapter_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

void main() {}


example/lib/main_zh.dart
import 'package:flutter/material.dart';
import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Screen Adapter Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    ScreenAdapter.init(context);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
          children: <Widget>[
            Text(
              '注：当前默认 750 * 1334 的设计稿尺寸',
              style: TextStyle(color: Colors.red),
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                Container(
                  width: ScreenAdapter.screenWidth / 2,
                  height: ScreenAdapter.value(200),
                  color: Colors.blue,
                  child: Column(
                    children: [
                      Text('w：${ScreenAdapter.screenWidthPx / 2} px'),
                      Text(
                        'w：${ScreenAdapter.screenWidth / 2} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                      Text(
                        'h：${ScreenAdapter.value(200)} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                    ],
                  ),
                ),
                Container(
                  width: ScreenAdapter.screenWidth / 2,
                  height: ScreenAdapter.value(200),
                  color: Colors.red,
                  child: Column(
                    children: [
                      Text('w：${ScreenAdapter.screenWidthPx / 2} px'),
                      Text(
                        'w：${ScreenAdapter.screenWidth / 2} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                      Text(
                        'h：${ScreenAdapter.value(200)} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                SizedBox(width: ScreenAdapter.value(8)),
                Container(
                  width: ScreenAdapter.value(100),
                  height: ScreenAdapter.value(100),
                  color: Colors.yellow,
                ),
                SizedBox(width: ScreenAdapter.value(8)),
                Column(
                  children: [
                    Text(
                        'ScreenAdapter.width(100)：${ScreenAdapter.value(100)}'),
                    Text(
                        'ScreenAdapter.height(100)：${ScreenAdapter.value(100)}'),
                  ],
                )
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                Container(
                  width: 375,
                  height: 100,
                  color: Colors.yellow,
                  child: Column(
                    children: [
                      Text('直接设置'),
                      Text('width: 375'),
                      Text('height: 100'),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Text('设备宽度： ${ScreenAdapter.screenWidthPx} px'),
            Text('设备高度： ${ScreenAdapter.screenHeightPx} px'),
            Text('设备宽度： ${ScreenAdapter.screenWidth} dp'),
            Text('设备高度： ${ScreenAdapter.screenHeight} dp'),
            Text('设备像素密度： ${ScreenAdapter.pixelRatio}'),
            Text('设备实际宽度 dp 与设计稿 px 比值： ${ScreenAdapter.scaleWidth}'),
            Text('设备实际高度 dp 与设计稿 px 比值： ${ScreenAdapter.scaleHeight}'),
            SizedBox(height: ScreenAdapter.value(16)),
            Text('字体缩放比例： ${ScreenAdapter.textScaleFactor}'),
            Text(
              '我是会缩放的文字',
              style: TextStyle(
                fontSize: ScreenAdapter.fontSize(24),
              ),
            ),
            Text(
              '我是不会缩放的文字',
              style: TextStyle(
                fontSize: 24,
              ),
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/screen_adapter.dart
import 'package:flutter/material.dart';

class ScreenAdapter {
  /// 设计稿尺寸
  ///
  /// 默认尺寸 `const Size(750.0, 1334.0)`
  static Size _designSize;

  /// 屏幕宽度
  ///
  static double _screenWidth;

  /// 屏幕高度
  ///
  static double _screenHeight;

  /// 设计稿宽度跟实际屏幕宽度比值
  ///
  static double _scaleWidth;

  /// 设计稿高度跟实际屏幕高度比值
  ///
  static double _scaleHeight;

  /// 设备物理像素
  ///
  static double _pixelRatio;

  /// 设备字体缩放因子
  ///
  static double _textScaleFactor;

  static init(BuildContext context, {Size designSize = const Size(750, 1334)}) {
    MediaQueryData mediaQuery = MediaQuery.of(context);
    _designSize = designSize;
    _screenWidth = mediaQuery.size.width;
    _screenHeight = mediaQuery.size.height;
    _scaleWidth = _screenWidth / _designSize.width;
    _scaleHeight = _screenHeight / _designSize.height;
    _pixelRatio = mediaQuery.devicePixelRatio;
    _textScaleFactor = mediaQuery.textScaleFactor;
  }

  /// 获取当前设备像素比
  ///
  static double get pixelRatio => _pixelRatio;

  /// 获取当前设备宽度与设计宽度比值
  ///
  static double get scaleWidth => _scaleWidth;

  /// 获取当前设备高度与设计高度比值
  ///
  static double get scaleHeight => _scaleHeight;

  /// 文字缩放比例
  ///
  static double get scaleText => _scaleWidth;

  /// 根据设备信息得到自适应值
  /// 用来使用不同屏幕达到跟设计稿一致效果
  ///
  /// [vertical] 是否是垂直自适应数值，默认是水平
  static double value(num value, {vertical = false}) =>
      vertical ? _scaleHeight * value : _scaleWidth * value;

  /// 文字适配
  ///
  static double fontSize(num fontSize) => fontSize * scaleText;

  /// 获取当前设备屏幕宽度
  ///
  static double get screenWidth => _screenWidth;

  /// 获取当前设备屏幕高度
  ///
  static double get screenHeight => _screenHeight;

  /// 获取当前设备文字缩放因子
  ///
  static double get textScaleFactor => _textScaleFactor;

  /// 获取当前设备屏幕宽度 px
  ///
  static double get screenWidthPx => _screenWidth * pixelRatio;

  /// 获取当前设备屏幕高度 px
  ///
  static double get screenHeightPx => _screenHeight * pixelRatio;
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_screen_adapter/flutter_screen_adapter.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Screen Adapter Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    ScreenAdapter.init(context);
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        child: Column(
          children: <Widget>[
            Text(
              'Note: The current default design draft size of 750 * 1334',
              style: TextStyle(color: Colors.red),
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                Container(
                  width: ScreenAdapter.screenWidth / 2,
                  height: ScreenAdapter.value(200),
                  color: Colors.blue,
                  child: Column(
                    children: [
                      Text('w：${ScreenAdapter.screenWidthPx / 2} px'),
                      Text(
                        'w：${ScreenAdapter.screenWidth / 2} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                      Text(
                        'h：${ScreenAdapter.value(200)} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                    ],
                  ),
                ),
                Container(
                  width: ScreenAdapter.screenWidth / 2,
                  height: ScreenAdapter.value(200),
                  color: Colors.red,
                  child: Column(
                    children: [
                      Text('w：${ScreenAdapter.screenWidthPx / 2} px'),
                      Text(
                        'w：${ScreenAdapter.screenWidth / 2} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                      Text(
                        'h：${ScreenAdapter.value(200)} dp',
                        style: TextStyle(color: Colors.white),
                      ),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                SizedBox(width: ScreenAdapter.value(8)),
                Container(
                  width: ScreenAdapter.value(100),
                  height: ScreenAdapter.value(100),
                  color: Colors.yellow,
                ),
                SizedBox(width: ScreenAdapter.value(8)),
                Column(
                  children: [
                    Text(
                        'ScreenAdapter.width(100)：${ScreenAdapter.value(100)}'),
                    Text(
                        'ScreenAdapter.height(100)：${ScreenAdapter.value(100)}'),
                  ],
                )
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Row(
              children: [
                Container(
                  width: 375,
                  height: 100,
                  color: Colors.yellow,
                  child: Column(
                    children: [
                      Text('Direct setting'),
                      Text('width: 375'),
                      Text('height: 100'),
                    ],
                  ),
                ),
              ],
            ),
            SizedBox(height: ScreenAdapter.value(8)),
            Text('Device width： ${ScreenAdapter.screenWidthPx} px'),
            Text('Device height： ${ScreenAdapter.screenHeightPx} px'),
            Text('Device width： ${ScreenAdapter.screenWidth} dp'),
            Text('Device height： ${ScreenAdapter.screenHeight} dp'),
            Text('Device pixel density： ${ScreenAdapter.pixelRatio}'),
            Text(
                'The ratio of the actual width of the device dp to the design draft px： ${ScreenAdapter.scaleWidth}'),
            Text(
                'The actual height of the equipment dp and the design draft px ratio： ${ScreenAdapter.scaleHeight}'),
            SizedBox(height: ScreenAdapter.value(16)),
            Text('Font scaling： ${ScreenAdapter.textScaleFactor}'),
            Text(
              'I am zoomable text',
              style: TextStyle(
                fontSize: ScreenAdapter.fontSize(24),
              ),
            ),
            Text(
              'I will not zoom the text',
              style: TextStyle(
                fontSize: 24,
              ),
            ),
          ],
        ),
      ),
    );
  }
}
